framework/src/media/audio/audio_manager.c

```diff
@@ -706,41 +723,52 @@ audio_manager_result_t set_audio_stream_in(unsigned int channels, unsigned int s
 
 	pthread_mutex_lock(&(card->card_mutex));
 
-	/* TODO Resample channel logic & sample rate based on policy should be added */
-	if (channels > AUDIO_STREAM_CHANNEL_MONO) {
-		channels = AUDIO_STREAM_CHANNEL_STEREO;
-	}
-
 	memset(&config, 0, sizeof(struct pcm_config));
-	config.channels = channels;
 	config.rate = get_closest_samprate(sample_rate, INPUT);
 	config.format = format;
 	config.period_size = AUDIO_STREAM_VOICE_RECOGNITION_PERIOD_SIZE;
 	config.period_count = AUDIO_STREAM_VOICE_RECOGNITION_PERIOD_COUNT;
-
-	medvdbg("[IN] Device samplerate: %u, User requested: %u\n", config.rate, sample_rate);
+	config.channels = channel_num;
+	medvdbg("[IN] Device samplerate: %u, User requested: %u   Device channel: %d\n", config.rate, sample_rate, channel_num);
 	card->pcm = pcm_open(g_actual_audio_in_card_id, card->device_id, PCM_IN, &config);
 	if (!pcm_is_ready(card->pcm)) {
 		meddbg("fail to pcm_is_ready() error : %s", pcm_get_error(card->pcm));
 		ret = AUDIO_MANAGER_CARD_NOT_READY;
 		goto error_with_pcm;
 	}
 
-	if (sample_rate == config.rate) {
-		card->resample.necessary = false;
-	} else {
-		card->resample.necessary = true;
-		card->resample.from = config.rate;
-		card->resample.to = sample_rate;
+	card->resample.format = pcm_format_to_bits(pcm_get_format(card->pcm)) >> 3;	// ToDo: Change into the argument "format".
+
+	card->resample.user_channel = channels;
+	rechanneling_ratio = (float)card->resample.user_channel / (float)config.channels;
+	if (rechanneling_ratio < 1) {
+		card->resample.rechanneling_buffer_size = get_user_input_frames_to_byte(get_input_frame_count()) / rechanneling_ratio;
+		card->resample.rechanneling_buffer = malloc(card->resample.rechanneling_buffer_size);
+		medvdbg("Rechanneling ratio = %f, rechanneling buffer size = %d\n", rechanneling_ratio, card->resample.rechanneling_buffer_size);
+	}
+
+	card->resample.ratio = 1;
+	if (sample_rate != config.rate) {
+		float resample_buffer_size = (float)get_input_frame_count();
+
+		card->resample.user_sample_rate = sample_rate;
 		card->resample.ratio = (float)sample_rate / (float)config.rate;
 		card->resample.handle = src_init(CONFIG_AUDIO_RESAMPLER_BUFSIZE);
-		card->resample.buffer = malloc((int)((float)get_input_frames_to_byte(get_input_frame_count()) / card->resample.ratio) + 1);	// +1 for floating point margin
+
+		resample_buffer_size /= card->resample.ratio;
+		if (resample_buffer_size - (int)resample_buffer_size > 0) {
+			resample_buffer_size = (int)resample_buffer_size + 1;
```

Comment:
please check my old comment regarding this and check resampler.
If resampler keep remained buffer in ring buffer, then we don't have to add +1,
If it doesn't keep it, then basic frame size(bit>>3 * channel) should be added.